utils/panel/panel.go

```diff
@@ -0,0 +1,173 @@
+package panel
+
+import (
+	"errors"
+
+	"github.com/mattermost/mattermost-plugin-api/utils/bot/logger"
+	"github.com/mattermost/mattermost-plugin-api/utils/bot/poster"
+	"github.com/mattermost/mattermost-plugin-api/utils/common"
+	"github.com/mattermost/mattermost-plugin-api/utils/panel/settings"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type Panel interface {
+	Set(userID, settingID string, value interface{}) error
+	Print(userID string)
+	ToPost(userID string) (*model.Post, error)
+	Clear(userID string) error
+	URL() string
+	GetSettingIDs() []string
+}
+
+type panel struct {
+	settings       map[string]settings.Setting
+	settingKeys    []string
+	poster         poster.Poster
+	logger         logger.Logger
+	store          PanelStore
+	settingHandler string
+	pluginURL      string
+}
+
+func NewSettingsPanel(settingList []settings.Setting, poster poster.Poster, logger logger.Logger, store PanelStore, settingHandler, pluginURL string) Panel {
+	settingsMap := make(map[string]settings.Setting)
+	settingKeys := []string{}
+	for _, s := range settingList {
+		settingsMap[s.GetID()] = s
+		settingKeys = append(settingKeys, s.GetID())
+	}
+
+	panel := &panel{
+		settings:       settingsMap,
+		settingKeys:    settingKeys,
+		poster:         poster,
+		logger:         logger,
+		store:          store,
+		settingHandler: settingHandler,
+		pluginURL:      pluginURL,
+	}
+
+	for _, s := range settingsMap {
+		ftf := s.GetFreetextFetcher()
+		if ftf == nil {
+			continue
+		}
+
+		ftf.UpdateHooks(nil, panel.ftOnFetch, panel.ftOnCancel)
+	}
+
+	return panel
+}
+
+func (p *panel) Set(userID, settingID string, value interface{}) error {
+	s, ok := p.settings[settingID]
+	if !ok {
+		return errors.New("cannot find setting " + settingID)
+	}
+
+	err := s.Set(userID, value)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (p *panel) GetSettingIDs() []string {
+	return p.settingKeys
+}
+
+func (p *panel) URL() string {
+	return p.settingHandler
+}
+
+func (p *panel) Print(userID string) {
+	err := p.cleanPreviousSettingsPosts(userID)
+	if err != nil {
+		p.logger.Errorf("could not clean previous setting post, " + err.Error())
+	}
+
+	sas := []*model.SlackAttachment{}
+	for _, key := range p.settingKeys {
+		s := p.settings[key]
+		sa, loopErr := s.GetSlackAttachments(userID, p.pluginURL+p.settingHandler, p.isSettingDisabled(userID, s))
+		if loopErr != nil {
+			p.logger.Errorf("error creating the slack attachment, err=" + loopErr.Error())
+			continue
+		}
+		sas = append(sas, sa)
+	}
+	postID, err := p.poster.DMWithAttachments(userID, sas...)
+	if err != nil {
+		p.logger.Errorf("error creating the message, err=", err.Error())
+		return
+	}
+
+	err = p.store.SetPanelPostID(userID, postID)
+	if err != nil {
+		p.logger.Errorf("could not set the post IDs, err=", err.Error())
+	}
+}
+
+func (p *panel) ToPost(userID string) (*model.Post, error) {
```

Comment:
Any way to write unit tests for these sort of functions?

Reply:
Once the code is "closed", I will invest some effort on commenting, and will open tickets for unit testing.